presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PartialAggregationPushDown.java

```diff
@@ -140,6 +158,147 @@ public PlanNode visitAggregation(AggregationNode node, RewriteContext<Void> cont
+        {
+            Optional<List<Symbol>> groupingSet = getPushedDownGroupingSet(node, child, ImmutableSet.copyOf(child.getRight().getOutputSymbols()));
+            if (!groupingSet.isPresent()) {
+                return context.defaultRewrite(node);
+            }
+
+            AggregationNode pushedAggregation = replaceAggregationSource(node, child.getRight(), child.getCriteria(), groupingSet.get(), context);
+            return pushPartialToJoin(pushedAggregation, child, context.rewrite(child.getLeft()), pushedAggregation, child.getLeft().getOutputSymbols());
+        }
+
+        private PlanNode pushPartialToJoin(
+                AggregationNode pushedAggregation,
+                JoinNode child,
+                PlanNode leftChild,
+                PlanNode rightChild,
+                Collection<Symbol> otherSymbols)
+        {
+            ImmutableList.Builder<Symbol> outputSymbols = ImmutableList.builder();
+            outputSymbols.addAll(pushedAggregation.getOutputSymbols());
+            outputSymbols.addAll(otherSymbols);
+
+            return new JoinNode(
+                    child.getId(),
+                    child.getType(),
+                    leftChild,
+                    rightChild,
+                    child.getCriteria(),
+                    outputSymbols.build(),
+                    child.getFilter(),
+                    child.getLeftHashSymbol(),
+                    child.getRightHashSymbol());
+        }
+
+        private AggregationNode replaceAggregationSource(AggregationNode aggregation, PlanNode source, List<EquiJoinClause> criteria, List<Symbol> groupingSet, RewriteContext<Void> context)
+        {
+            PlanNode rewrittenSource = context.rewrite(source);
+            ImmutableSet<Symbol> rewrittenSourceSymbols = ImmutableSet.copyOf(rewrittenSource.getOutputSymbols());
+            ImmutableMap.Builder<Symbol, Symbol> mapping = ImmutableMap.builder();
+
+            for (EquiJoinClause joinClause : criteria) {
+                if (rewrittenSourceSymbols.contains(joinClause.getLeft())) {
+                    mapping.put(joinClause.getRight(), joinClause.getLeft());
+                }
+                else {
+                    mapping.put(joinClause.getLeft(), joinClause.getRight());
+                }
+            }
+
+            AggregationNode pushedAggregation = new AggregationNode(
+                    aggregation.getId(),
+                    aggregation.getSource(),
+                    aggregation.getAggregations(),
+                    aggregation.getFunctions(),
+                    aggregation.getMasks(),
+                    ImmutableList.of(groupingSet),
+                    aggregation.getStep(),
+                    aggregation.getHashSymbol(),
+                    aggregation.getGroupIdSymbol());
+            return new SymbolMapper(mapping.build()).map(pushedAggregation, source);
```

Comment:
To make sure I understand:  The idea behind this mapping here is that anything in the grouping set that referenced  e.g. the left side of the join clause if we're pushing right gets rewritten to be about the right side. And we can do that because they have an equality predicate between them.